app/lib/shared/Core/entities/DocumentEntity/mandatoryTest_6_1_3.js

```diff
@@ -0,0 +1,90 @@
+/**
+ * @typedef {object} FullProductName
+ * @property {unknown} product_id
+ */
+
+/**
+ * @typedef {object} Relationship
+ * @property {unknown} relates_to_product_reference
+ * @property {unknown} product_reference
+ * @property {FullProductName | null} [full_product_name]
+ */
+
+/**
+ * @param {any} doc
+ */
+export default function (doc) {
+  /** @type {Array<{ message: string; instancePath: string }>} */
+  const errors = []
+  let isValid = true
+
+  if (!Array.isArray(doc.product_tree?.relationships)) {
+    return { isValid, errors }
+  }
+
+  /** @type {Array<Relationship>} */
+  const relationships = doc.product_tree.relationships
+
+  relationships.forEach((_, index) => {
+    /** @type {Set<string>} */
+    const erroredAttributeSet = new Set()
+    search([], relationships, index, null, ({ key }) => {
+      if (typeof key === 'string' && !erroredAttributeSet.has(key)) {
+        erroredAttributeSet.add(key)
```

Comment:
It prevents that errors for the same field are recorded multiple times. Since we check the attributes in a serial fashion there are "multiple entrypoints" to the same cycle which leads to duplicate error messages.